Replace project_bootstrap script with playbook

jcpowermac opened the pull request:
- Grab user oauth token from OpenShift API
- Move host_vars to group_vars
- Create new host_vars specifically openshift_{username,password}
- Adds the ability to use ansible-vault on new host_vars

codificat commented on ansible/group_vars/all.yml.example:
```diff
@@ -34,4 +35,4 @@ notify_email_list: aweiteka@redhat.com
 # --------------------------------
 # the token/clusterhost/project_name are set in host_vars for each environment
-oc: oc --token={{ token }} --server=https://{{ clusterhost }} -n {{ project_name }}
+oc: oc --insecure-skip-tls-verify --token={{ token }} --server=https://{{ clusterhost }} -n {{ project_name }}
```
It's not clear to me why/if we should add `insecure-skip-tls-verify` by default... maybe add a comment instead mentioning that this is the place to add it if (absolutely) necessary?

codificat commented on ansible/group_vars/dev.yml.example:
```diff
@@ -1,6 +1,8 @@
+---
 project_name: dev
+project_display_name: "Software lifecycle"
+project_description: "Software Development Lifecycle (SDLC)"
```
Good idea to add these vars to allow project display/description!

Maybe take the opportunity to customise them slightly? E.g. "Development project" vs "Staging project" or something along those lines...

codificat commented on ansible/group_vars/prod.yml.example:
```diff
@@ -1,6 +1,8 @@
+---
 project_name: prod
+project_display_name: "Software lifecycle"
+project_description: "Software Development Lifecycle (SDLC)"
```
Ditto on potentially setting a display/description specific to this project

codificat commented on ansible/group_vars/registry.yml.example:
```diff
@@ -1,6 +1,8 @@
+---
 project_name: registry
+project_display_name: "Software lifecycle"
+project_description: "Software Development Lifecycle (SDLC)"
```
Ditto on potentially setting a display/description specific to this project

codificat commented on ansible/group_vars/stage.yml.example:
```diff
@@ -1,6 +1,8 @@
+---
 project_name: stage
+project_display_name: "Software lifecycle"
+project_description: "Software Development Lifecycle (SDLC)"
```
Ditto on potentially setting a display/description specific to this project

codificat commented on ansible/inventory.yml:
```diff
@@ -1,13 +1,17 @@
+        prod-1
     registry:
+      hosts:
+        registry-1
```
Just wondering... why the addition of `-1` to the hosts here? (and related changes)

Is it about potentially handling multiple dev/stg/prod clusters? I fail to see the use case here though...

jcpowermac replied:
@codificat I moved the host_vars to a group_vars.  I thought it would be confusing if the group and hosts have the same name...I don't think you can do it either.  I also wanted the ability to use `ansible-vault` on the `openshift_username` and `openshift_password`.

Example: https://github.com/jcpowermac/jenkins-on-openshift/tree/ansible_updates_testing/ansible/host_vars

codificat commented on ansible/main.yml:
```diff
@@ -1,4 +1,41 @@
+        url: "https://{{ clusterhost }}/oauth/authorize?response_type=token&client_id=openshift-challenging-client"
+        user: "{{ openshift_username }}"
+        password: "{{ openshift_password }}"
+        validate_certs: False
```
I'd vote against disabling cert validation by default

jcpowermac replied:
@codificat will add a variable for self-signed certs for both this and the insecure flag.

codificat commented on ansible/main.yml:
```diff
@@ -1,4 +1,41 @@
+        password: "{{ openshift_password }}"
+        validate_certs: False
+        force_basic_auth: True
+      register: login
```
I think this would not work on clusters whose identity providers don't support challenge authentication... and in any case we probably want to support pre-authentication...

jcpowermac replied:
@codificat ok I will add a when clause if token is already defined.

codificat reviewed (changes requested):
Thanks for the changes, I believe it's better to handle this from ansible.

Some comments/questions, and something I believe we should fix re. authentication

codificat commented:
By the way, `ansible/README.md` should also be updated as part of this PR

codificat commented on ansible/README.md:
```diff
@@ -2,32 +2,35 @@
-        oc policy add-role-to-user registry-admin -n $PROJECT -z ansible
+2. Using the 'host_vars/[environment]-1.yml.example' files, rename each file to remove '.example'.  Depending on your authentication method to OpenShift either use:
+  - `openshift_username`
+  - `openshift_password`
```
Being a bit picky here but... I think this could be a bit more clear/verbose, clarifying "either use username+password OR token", and possibly mention why (e.g. if your auth mechanism requires login to an external auth provider) and how (`oc whoami -t`) you would use token instead of user/pwd.

It could be good to reference anisble-vault too given that your change enables its use...

codificat reviewed (commented):
Just a minor comment about doc improvement.

This also needs a rebase.

Other than that, LGTM!

jcpowermac committed:
Replace project_bootstrap with playbook update

- Grab user oauth token from OpenShift API or use a token if defined
- In bootstrap do not use kube config
- Move host_vars to group_vars
- Create new host_vars specifically openshift_{username,password}
- Adds the ability to use ansible-vault on new host_vars
- Update README.md

jcpowermac committed:
formatting
